CMakeLists.txt

```diff
@@ -102,62 +144,227 @@ if(LINK_WITH_FIR)
+
+if (NOT(FLANG_DEFAULT_RTLIB STREQUAL ""))
+  message(WARNING "Resetting Flang's default runtime library to use platform default.")
+  set(FLANG_DEFAULT_RTLIB "" CACHE STRING
+      "Default runtime library to use (empty for platform default)" FORCE)
+endif()
+
+
+
+set(PACKAGE_VERSION "${LLVM_PACKAGE_VERSION}")
+# Override LLVM versioning for now...
 set(FLANG_VERSION_MAJOR      "0")
 set(FLANG_VERSION_MINOR      "1")
 set(FLANG_VERSION_PATCHLEVEL "0")
+
+
+if (NOT DEFINED FLANG_VERSION_MAJOR)
+  set(FLANG_VERSION_MAJOR ${LLVM_VERSION_MAJOR})
+endif()
+
+if (NOT DEFINED FLANG_VERSION_MINOR)
+  set(FLANG_VERSION_MINOR ${LLVM_VERSION_MINOR})
+endif()
+
+if (NOT DEFINED FLANG_VERSION_PATCHLEVEL)
+  set(FLANG_VERSION_PATCHLEVEL ${LLVM_VERSION_PATCH})
+endif()
+
+# Unlike PACKAGE_VERSION, FLANG_VERSION does not include LLVM_VERSION_SUFFIX.
 set(FLANG_VERSION "${FLANG_VERSION_MAJOR}.${FLANG_VERSION_MINOR}.${FLANG_VERSION_PATCHLEVEL}")
-message(STATUS "FLANG version: ${FLANG_VERSION}")
+message(STATUS "Flang version: ${FLANG_VERSION}")
+# Flang executable version information
+set(FLANG_EXECUTABLE_VERSION
+    "${FLANG_VERSION_MAJOR}" CACHE STRING
+    "Major version number to appended to the flang executable name.")
+set(LIBFLANG_LIBRARY_VERSION
+    "${FLANG_VERSION_MAJOR}" CACHE STRING
+    "Major version number to appended to the libflang library.")
 
-set(FLANG_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
-set(FLANG_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR})
-set(FLANG_TOOLS_DIR  ${FLANG_BINARY_DIR}/tools/f18/bin)
+mark_as_advanced(FLANG_EXECUTABLE_VERSION LIBFLANG_LIBRARY_VERSION)
 
-include_directories(BEFORE
-  ${FLANG_BINARY_DIR}/include
-  ${FLANG_SOURCE_DIR}/include
-  )
+set(FLANG_VENDOR ${PACKAGE_VENDOR} CACHE STRING
+  "Vendor-specific Flang version information.")
+set(FLANG_VENDOR_UTI "org.llvm.flang" CACHE STRING
+  "Vendor-specific uti.")
 
-enable_testing()
+if (FLANG_VENDOR)
+  add_definitions(-DFLANG_VENDOR="${FLANG_VENDOR} ")
+endif()
 
-add_subdirectory(include/flang)
-add_subdirectory(lib)
-add_subdirectory(runtime)
-add_subdirectory(test)
-add_subdirectory(tools)
-add_subdirectory(test-lit)
+set(FLANG_REPOSITORY_STRING "" CACHE STRING
+  "Vendor-specific text for showing the repository the source is taken from.")
+if (FLANG_REPOSITORY_STRING)
+  add_definitions(-DFLANG_REPOSITORY_STRING="${FLANG_REPOSITORY_STRING}")
+endif()
 
+# Configure Flang's Version.inc file.
+configure_file(
+  ${CMAKE_CURRENT_SOURCE_DIR}/include/flang/Version.inc.in
+  ${CMAKE_CURRENT_BINARY_DIR}/include/flang/Version.inc)
+# Configure Flang's version info header file.
 configure_file(
   ${FLANG_SOURCE_DIR}/include/flang/Config/config.h.cmake
   ${FLANG_BINARY_DIR}/include/flang/Config/config.h)
+
+# Add global F18 flags.
+set(CMAKE_CXX_FLAGS "-fno-rtti -fno-exceptions -pedantic -Wall -Wextra -Werror -Wcast-qual -Wimplicit-fallthrough -Wdelete-non-virtual-dtor ${CMAKE_CXX_FLAGS}")
```

Comment:
Is it possible to add checks for these flags before using them?
Also, is `-fno-rtti` still needed? `functional` header cannot be included in Windows with this flag.

Reply:
I'll need to look in to what LLVM does here. We do need -fno-rtti as llvm builds with that, at least on Linux, macOS and *BSD. I don't suppose you could send a list of what flags LLVM uses by default on Windows?

Reply:
I think it might be. My build environment was tripping over some RTTI link failures without it. YMMV.

Reply:
I'm cross compiling at the moment, so lots of flags are added from my build environment.
LLVM's RTTI handling is at https://github.com/llvm/llvm-project/blob/50a6d3a6486d81d21a2c31ce8522321e19bed35e/llvm/cmake/modules/AddLLVM.cmake#L38-L53. Exception handling https://github.com/llvm/llvm-project/blob/50a6d3a6486d81d21a2c31ce8522321e19bed35e/llvm/cmake/modules/AddLLVM.cmake#L13-L36

Reply:
Would it be OK with you if we merged this now and opened another issue/PR to fix windows build issues? I'll find out if I can set up a native Windows build environment at some point to debug these issues.

Reply:
Sure